### PR TITLE
Many2Many through pivot table support

### PR DESCRIPTION
**Pull request by realglebivanov**
Hi Michael! I can write some tests a bit later - have another work for now.

### CODE REVIEW

**Comment by michaeldyrynda** [comment]
Hey @glebivanov816 - thanks for this. Are you able to add some integration tests to cover this case just to make sure it's not broken in future?

**Comment by michaeldyrynda** [comment]
Just started reviewing :) <br><br>:octocat: *Sent from [GH](http://nodegh.io).*

File: tests/CascadeSoftDeletesIntegrationTest.php [hunk 1]
```diff
@@ -68,6 +83,26 @@ public function it_cascades_deletes_when_deleting_a_parent_model()
         $this->assertCount(0, Tests\Entities\Comment::where('post_id', $post->id)->get());
     }
 
+    /* Many2Many with pivot table(deletes entries in pivot) */
+    public function it_cascades_deletes_entries_from_pivot_table()
```
Comment (michaeldyrynda) [inline tests/CascadeSoftDeletesIntegrationTest.php]: This test doesn't run without the `/** @test */` comment notation before it

File: tests/CascadeSoftDeletesIntegrationTest.php [hunk 2]
```diff
@@ -68,6 +83,25 @@ public function it_cascades_deletes_when_deleting_a_parent_model()
...
+    {
+        $author = Tests\Entities\Author::create(['name' => 'ManyToManyTestAuthor']);
+
+        $this->attachPostTypesToAuthor($author);
+        $this->assertCount(2, $author->posttypes);
+
+        $author->delete();
+
+        $manager = Manager::$instance;
```
Comment (michaeldyrynda) [inline tests/CascadeSoftDeletesIntegrationTest.php]: `$instance` is a protected property and can't be accessed like this. I'm not sure what you're trying to achieve here.

File: src/CascadeSoftDeletes.php [hunk 3]
```diff
@@ -41,7 +41,12 @@ protected static function bootCascadeSoftDeletes()
                     $model->{$relationship}->{$delete}();
                 } else {
                     foreach ($model->{$relationship} as $child) {
-                        $child->{$delete}();
+                        if ($child->pivot){
+                            $child->pivot->delete();
```
Comment (michaeldyrynda) [inline src/CascadeSoftDeletes.php]: You should use the `$delete` variable here, in order to ensure you perform a `forceDelete` if that is what the user is running.

`$child->pivot->{$method}();`

**Comment by michaeldyrynda** [comment]
Hey @glebivanov816 - I've added some comments for you to take a look at. If you could also review and address the issues that nitpick identified.

I don't think this is going to work in any event; I don't think you can call `delete()` on the `pivot` attribute in this way, as there's other behaviour that ought to happen on a many to many?

It may be better to check the relation type and call `detach` on it (probably passing an array of all the IDs, so as to do it in one query).

**Comment by glebivanov816** [comment]
Everything is ok (I mean, there's no problem to fix the test), but the script itself. There's more about questioning to database inside a loop, rather than deleting with pivot->delete(). Pivot returns you an instance of Illuminate\Database\Eloquent\Pivot which does have only method delete(), which deletes concrete entry from a pivot table.

But to not loop $model->{$relation} we can really just detach() with no arguments - it works both with one-to-many and many-to-many.

**Comment by glebivanov816** [comment]
Everything is ok (I mean, there's no problem to fix the test), but the script itself. There's more about questioning to database inside a loop, rather than deleting with pivot->delete(). Pivot returns you an instance of Illuminate\Database\Eloquent\Pivot which does have only method delete(), which deletes concrete entry from a pivot table.

Since pivot table is not a true entity, it cannot handle softdeletes. When we use detach(), i will destroy entries in pivot table. But $child is still querying to database in loop.

**Comment by michaeldyrynda** [comment]
Just started reviewing :) <br><br>:octocat: *Sent from [GH](http://nodegh.io).*

**Comment by michaeldyrynda** [comment]
Hey @glebivanov816 - I've spent a bit of time on this tonight and got it to a state where the code is reasonably clean and making all the tests pass (your new one for pivots included).

If I can get you to try out the code in the [pr-13 branch](https://github.com/michaeldyrynda/laravel-cascade-soft-deletes/tree/pr-13) and see that it works for your use case, that'd be great.

Keep in mind, that as your dataset grows, this is going to become less and less performant and it'll really time to start moving away from soft deletes and into database-level (cascading) deletes.

**Comment by endever** [comment]
Hey @michaeldyrynda . I see this guy has been hanging around for a while, what are the chances of merging this or supporting cascade delete/detach of the many to many joining table, instead of the related model?

**Comment by pierrocknroll** [comment]
Please merge ! :)

**Pull request closed without merging**